### DISCUSSION TITLE
Shortest path distances of all node pairs between source and target

### DISCUSSION THREAD

**Discussion by YujiKopi** [networkx/networkx#5268] [Q&A] [answered]
Hello,

I was wondering if there is any equivalent function to get the shortest path distances of all node pairs between a source and a target as MATLAB distances function (https://fr.mathworks.com/help/matlab/ref/graph.distances.html) ?

Thanks a lot for your consideration.

**Comment by MridulS** [comment, marked as answer]
[all_pairs_shortest_path_length](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.unweighted.all_pairs_shortest_path_length.html#networkx.algorithms.shortest_paths.unweighted.all_pairs_shortest_path_length) will find the distances between all the pairs of nodes.

If it's a weighted network, then [all_pairs_dijkstra_path_length](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.weighted.all_pairs_dijkstra_path_length.html#networkx.algorithms.shortest_paths.weighted.all_pairs_dijkstra_path_length) can be used :)

**Reply by YujiKopi** [reply]
Yes thanks a lot, but I'm way more interested by the source and target functionality as _**distances(G,s,t) restricts the target nodes to the nodes defined by t, such that d(i,j) is the distance from node s(i) to node t(j)**_  instead of just  **_distances(G)_** as [all_pairs_shortest_path_length](url)  can do (Cause yes this is unweighted). 

Is there a way ? But many thanks !

**Reply by dschult** [reply]
For unweighted graphs the most efficient way to get all shortest paths from nodes in s to nodes in t is **usually** to use a loop over nodes in s with the function `single_source_shortest_path_length`.  The only time that isn't true is when the number of nodes in t is small or all sources and targets happen to be arranged in a local region within the graph.  For those cases, you should use `shortest_path_length` looping over each of the source and each of the target nodes. There is no easy way to tell ahead of time, or during the process, which method will be faster.

So, we don't provide a function to compute all distances between each member of a set of nodes to each member of another set of nodes. I personally suggest `all_pairs_shortest_path_length`. But your case may be one of the few which performs better with the loops.  Keep in mind this is true for the Matlab code too.  They don't know which way to proceed and probably use some heuristic to choose whether to do a bunch of source - to all targets. Looking at their code, it seems that their inner level search function only takes source as an argument and checks that it is a valid node. So I suspect they are doing what I suggest at the top of this message.

**Reply by YujiKopi** [reply]
Ok many thanks, one last thing why is the length of **nx.single_source_shortest_path_length(Gp, source)** is less than number of all Graph nodes ?

**Reply by MridulS** [reply]
By length of `nx.single_source_shortest_path_length(Gp, source)` do you mean the length of the dictionary output of the function or the underlying length of path between source and every other node?

The output of `nx.single_source_shortest_path_length(Gp, source)`  should have an equal length to the number of nodes, example:
``` python
In [5]: G = nx.erdos_renyi_graph(10, 0.4)

In [6]: nx.single_source_shortest_path_length(G, 1)
Out[6]: {1: 0, 0: 1, 5: 1, 6: 1, 7: 1, 3: 2, 8: 2, 9: 2, 2: 3, 4: 3}

In [7]: len(nx.single_source_shortest_path_length(G, 1)) # Length of  nx.single_source_shortest_path_length
Out[7]: 10

In [8]: len(G) # Number of nodes in the Graph
Out[8]: 10
```

**Reply by YujiKopi** [reply]
![image](https://user-images.githubusercontent.com/97978392/150353135-494f4453-1518-4999-be39-e4f1a1893e2b.png)

Hmm I should miss something

**Reply by MridulS** [reply]
Ahhhhh yeah. That means you have some nodes which are unreachable from the node 1927. There is no path between node 1927 and the missing nodes.

**Reply by YujiKopi** [reply]
Ok ... Well none of my nodes can reach the 16 missings, quite strange, but thanks a lot !
